plugin/src/main/kotlin/com/karumi/hagu/plugin/source/SourceRoots.kt

```diff
@@ -0,0 +1,62 @@
+package com.karumi.hagu.plugin.source
+
+import com.karumi.hagu.plugin.HaguTask
+import org.gradle.api.Project
+import org.gradle.api.file.SourceDirectorySet
+import org.gradle.api.tasks.SourceSetContainer
+import org.gradle.api.tasks.TaskProvider
+import org.jetbrains.kotlin.gradle.dsl.KotlinMultiplatformExtension
+import org.jetbrains.kotlin.gradle.plugin.KotlinPlatformType
+import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinJvmAndroidCompilation
+import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
+import org.jetbrains.kotlin.konan.target.KonanTarget
+
+internal fun sources(project: Project): List<Source> {
+  project.extensions.findByType(KotlinMultiplatformExtension::class.java)?.let {
+    return it.sources()
+  }
+
+  val sourceSets = project.property("sourceSets") as SourceSetContainer
+  return listOf(Source(
+    type = KotlinPlatformType.jvm,
+    name = "main",
+    sourceSets = listOf("main"),
+    sourceDirectorySet = sourceSets.getByName("main").kotlin!!,
+    registerTaskDependency = { task ->
+      project.tasks.named("compileKotlin").configure { it.dependsOn(task) }
+    }
+  ))
+}
+
+private fun KotlinMultiplatformExtension.sources(): List<Source> =
+  targets.flatMap { target ->
+    return@flatMap target.compilations.mapNotNull { compilation ->
+      if (compilation.name.endsWith(suffix = "Test", ignoreCase = true)) {
+        return@mapNotNull null
+      }
+      Source(
+        type = target.platformType,
+        konanTarget = (target as? KotlinNativeTarget)?.konanTarget,
+        name = "${target.name}${compilation.name.capitalize()}",
+        variantName = (compilation as? KotlinJvmAndroidCompilation)?.name,
+        sourceDirectorySet = compilation.defaultSourceSet.kotlin,
+        sourceSets = compilation.allKotlinSourceSets.map { it.name },
+        registerTaskDependency = { task ->
+          (target as? KotlinNativeTarget)?.binaries?.forEach {
+            it.linkTask.dependsOn(task)
+          }
+          compilation.compileKotlinTask.dependsOn(task)
+        }
+      )
+    }
```

Comment:
I'm seeing too much nesting with maps and flatmaps. We can maybe extract a function that gets the `Source` instance from a compilation and leave it like:

```kotlin
targets.flatMap { target ->
    target.compilations.mapNotNull { it.toSource() } 
}
```

Reply:
Yes, I can improve it!